<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from erp42_gazebo.xacro             | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="erp42">
  <!-- ======================================== Variables ======================================== -->
  <!-- ======================================== Links ======================================== -->
  <!-- Base Footprint -->
  <link name="base_footprint"/>
  <!-- Base link -->
  <link name="base_link">
    <!-- Inertia -->
    <inertial>
      <mass value="160"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="20.43" ixy="0.584" ixz="-0.191" iyy="54.197" iyz="-0.2449" izz="72.2588"/>
    </inertial>
    <!-- Chassis -->
    <collision>
      <origin rpy="0 0 0" xyz="0.02 0 0"/>
      <geometry>
        <box size="1.580 0.950 0.316"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/mingq/workspace/ros2/erp42_ros/install/erp42_description/share/erp42_description/meshes/erp42_chassis.dae" scale="0.001 0.001 0.001"/>
      </geometry>
    </visual>
    <!-- Front bumper -->
    <collision>
      <origin rpy="0 0 0" xyz="0.9199999999999999 0 -0.075"/>
      <geometry>
        <box size="0.22 1.0 0.15"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 0" xyz="0.9199999999999999 0 -0.075"/>
      <geometry>
        <mesh filename="file:///home/mingq/workspace/ros2/erp42_ros/install/erp42_description/share/erp42_description/meshes/erp42_bumper.dae" scale="0.001 0.001 0.001"/>
      </geometry>
    </visual>
    <!-- Rear bumper -->
    <collision>
      <origin rpy="0 0 3.14159" xyz="-0.88 0 -0.075"/>
      <geometry>
        <box size="0.22 1.0 0.15"/>
      </geometry>
    </collision>
    <visual>
      <origin rpy="0 0 3.14159" xyz="-0.88 0 -0.075"/>
      <geometry>
        <mesh filename="file:///home/mingq/workspace/ros2/erp42_ros/install/erp42_description/share/erp42_description/meshes/erp42_bumper.dae" scale="0.001 0.001 0.001"/>
      </geometry>
    </visual>
  </link>
  <!-- Steering assemblies -->
  <link name="front_left_steering_hub">
    <!-- Inertia -->
    <inertial>
      <mass value="0.1"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
  </link>
  <link name="front_right_steering_hub">
    <!-- Inertia -->
    <inertial>
      <mass value="0.1"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
  </link>
  <link name="front_left_wheel">
    <!-- Inertia -->
    <inertial>
      <mass value="7.0"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.145425" ixy="0.0" ixz="0.0" iyy="0.25515" iyz="0.0" izz="0.145425"/>
    </inertial>
    <!-- Collision -->
    <collision>
      <origin rpy="-1.570795 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.175" radius="0.27"/>
      </geometry>
    </collision>
    <!-- Mesh files -->
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/mingq/workspace/ros2/erp42_ros/install/erp42_description/share/erp42_description/meshes/erp42_tire.dae" scale="0.001 0.001 0.001"/>
      </geometry>
    </visual>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/mingq/workspace/ros2/erp42_ros/install/erp42_description/share/erp42_description/meshes/erp42_rim.dae" scale="0.001 0.001 0.001"/>
      </geometry>
    </visual>
  </link>
  <link name="front_right_wheel">
    <!-- Inertia -->
    <inertial>
      <mass value="7.0"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.145425" ixy="0.0" ixz="0.0" iyy="0.25515" iyz="0.0" izz="0.145425"/>
    </inertial>
    <!-- Collision -->
    <collision>
      <origin rpy="-1.570795 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.175" radius="0.27"/>
      </geometry>
    </collision>
    <!-- Mesh files -->
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/mingq/workspace/ros2/erp42_ros/install/erp42_description/share/erp42_description/meshes/erp42_tire.dae" scale="0.001 0.001 0.001"/>
      </geometry>
    </visual>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/mingq/workspace/ros2/erp42_ros/install/erp42_description/share/erp42_description/meshes/erp42_rim.dae" scale="0.001 0.001 0.001"/>
      </geometry>
    </visual>
  </link>
  <link name="rear_left_wheel">
    <!-- Inertia -->
    <inertial>
      <mass value="7.0"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.145425" ixy="0.0" ixz="0.0" iyy="0.25515" iyz="0.0" izz="0.145425"/>
    </inertial>
    <!-- Collision -->
    <collision>
      <origin rpy="-1.570795 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.175" radius="0.27"/>
      </geometry>
    </collision>
    <!-- Mesh files -->
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/mingq/workspace/ros2/erp42_ros/install/erp42_description/share/erp42_description/meshes/erp42_tire.dae" scale="0.001 0.001 0.001"/>
      </geometry>
    </visual>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/mingq/workspace/ros2/erp42_ros/install/erp42_description/share/erp42_description/meshes/erp42_rim.dae" scale="0.001 0.001 0.001"/>
      </geometry>
    </visual>
  </link>
  <link name="rear_right_wheel">
    <!-- Inertia -->
    <inertial>
      <mass value="7.0"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.145425" ixy="0.0" ixz="0.0" iyy="0.25515" iyz="0.0" izz="0.145425"/>
    </inertial>
    <!-- Collision -->
    <collision>
      <origin rpy="-1.570795 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.175" radius="0.27"/>
      </geometry>
    </collision>
    <!-- Mesh files -->
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/mingq/workspace/ros2/erp42_ros/install/erp42_description/share/erp42_description/meshes/erp42_tire.dae" scale="0.001 0.001 0.001"/>
      </geometry>
    </visual>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="file:///home/mingq/workspace/ros2/erp42_ros/install/erp42_description/share/erp42_description/meshes/erp42_rim.dae" scale="0.001 0.001 0.001"/>
      </geometry>
    </visual>
  </link>
  <!-- ======================================== Joints ======================================== -->
  <!-- base_footprint to base_link -->
  <joint name="base_link_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0.27"/>
    <parent link="base_footprint"/>
    <child link="base_link"/>
  </joint>
  <!-- base_link to front_left_steering_hub -->
  <joint name="front_left_steering_hub_joint" type="revolute">
    <origin rpy="0 0 0" xyz="0.52 0.4925 0"/>
    <parent link="base_link"/>
    <child link="front_left_steering_hub"/>
    <axis xyz="0 0 1"/>
    <limit effort="100" lower="-0.698132" upper="0.698132" velocity="100"/>
    <dynamics damping="0.1" friction="0.1"/>
  </joint>
  <!-- base_link to front_right_steering_hub -->
  <joint name="front_right_steering_hub_joint" type="revolute">
    <origin rpy="0 0 0" xyz="0.52 -0.4925 0"/>
    <parent link="base_link"/>
    <child link="front_right_steering_hub"/>
    <axis xyz="0 0 1"/>
    <limit effort="100" lower="-0.698132" upper="0.698132" velocity="100"/>
    <dynamics damping="0.1" friction="0.1"/>
  </joint>
  <!-- front_left_steering_hub to front_right_wheel -->
  <joint name="front_left_wheel_joint" type="continuous">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="front_left_steering_hub"/>
    <child link="front_left_wheel"/>
    <axis xyz="0 1 0"/>
  </joint>
  <!-- front_right_steering_hub to front_right_wheel -->
  <joint name="front_right_wheel_joint" type="continuous">
    <origin rpy="3.14159 0 0" xyz="0 0 0"/>
    <parent link="front_right_steering_hub"/>
    <child link="front_right_wheel"/>
    <axis xyz="0 -1 0"/>
  </joint>
  <!-- base_link to rear_left_wheel -->
  <joint name="rear_left_wheel_joint" type="continuous">
    <origin rpy="0 0 0" xyz="-0.52 0.4925 0"/>
    <parent link="base_link"/>
    <child link="rear_left_wheel"/>
    <axis xyz="0 1 0"/>
  </joint>
  <!-- base_link to rear_right_wheel -->
  <joint name="rear_right_wheel_joint" type="continuous">
    <origin rpy="3.14159 0 0" xyz="-0.52 -0.4925 0"/>
    <parent link="base_link"/>
    <child link="rear_right_wheel"/>
    <axis xyz="0 -1 0"/>
  </joint>
  <!-- Friction coefficients -->
  <gazebo reference="front_left_wheel">
    <mu1>0.9</mu1>
    <mu2>0.9</mu2>
  </gazebo>
  <gazebo reference="front_right_wheel">
    <mu1>0.9</mu1>
    <mu2>0.9</mu2>
  </gazebo>
  <gazebo reference="rear_left_wheel">
    <mu1>0.9</mu1>
    <mu2>0.9</mu2>
  </gazebo>
  <gazebo reference="rear_right_wheel">
    <mu1>0.9</mu1>
    <mu2>0.9</mu2>
  </gazebo>
  <!-- Gazebo plugins -->
  <gazebo>
    <!-- Joint state publisher -->
    <plugin filename="libgazebo_ros_joint_state_publisher.so" name="joint_state_publisher">
      <!-- ROS parameters -->
      <ros>
        <namespace>erp42</namespace>
      </ros>
      <!-- Update rate -->
      <update_rate>100.0</update_rate>
      <!-- Joint names -->
      <joint_name>front_left_wheel_joint</joint_name>
      <joint_name>front_right_wheel_joint</joint_name>
      <joint_name>rear_left_wheel_joint</joint_name>
      <joint_name>rear_right_wheel_joint</joint_name>
      <joint_name>front_left_steering_hub_joint</joint_name>
      <joint_name>front_right_steering_hub_joint</joint_name>
    </plugin>
    <!-- Ackermann drive controller -->
    <plugin filename="libgazebo_ros_ackermann_drive.so" name="gazebo_ros_ackermann_drive">
      <!-- ROS parameters -->
      <ros>
        <namespace>erp42</namespace>
        <remapping>cmd_vel:=cmd_vel</remapping>
        <remapping>odom:=odometry_ground_truth</remapping>
      </ros>
      <!-- Update rate -->
      <update_rate>100.0</update_rate>
      <!-- Joint names -->
      <front_left_joint>front_left_wheel_joint</front_left_joint>
      <front_right_joint>front_right_wheel_joint</front_right_joint>
      <rear_left_joint>rear_left_wheel_joint</rear_left_joint>
      <rear_right_joint>rear_right_wheel_joint</rear_right_joint>
      <left_steering_joint>front_left_steering_hub_joint</left_steering_joint>
      <right_steering_joint>front_right_steering_hub_joint</right_steering_joint>
      <!-- Vehicle kinematics & dynamics parameters -->
      <max_steer>0.436332</max_steer>
      <max_speed>7.0</max_speed>
      <!-- Vehicle PID gains -->
      <left_steering_pid_gain>1000 0 200</left_steering_pid_gain>
      <left_steering_i_range>0 0</left_steering_i_range>
      <right_steering_pid_gain>1000 0 200</right_steering_pid_gain>
      <right_steering_i_range>0 0</right_steering_i_range>
      <linear_velocity_pid_gain>50 0 1</linear_velocity_pid_gain>
      <linear_velocity_i_range>0 0</linear_velocity_i_range>
      <!-- Odometry parameters -->
      <publish_odom>True</publish_odom>
      <publish_odom_tf>True</publish_odom_tf>
      <odometry_frame>odom</odometry_frame>
      <robot_base_frame>base_footprint</robot_base_frame>
    </plugin>
  </gazebo>
</robot>
